ee/server/service/mdm.go

```diff
@@ -1195,3 +1179,194 @@ func (svc *Service) GetMDMManualEnrollmentProfile(ctx context.Context) ([]byte,
+	if err := svc.authz.Authorize(ctx, &fleet.AppleBM{}, fleet.ActionWrite); err != nil {
+		return err
+	}
+
+	if err := svc.ds.DeleteABMToken(ctx, tokenID); err != nil {
+		return ctxerr.Wrap(ctx, err, "removing ABM token")
+	}
+
+	count, err := svc.ds.GetABMTokenCount(ctx)
+	if err != nil {
+		return ctxerr.Wrap(ctx, err, "getting ABM token count")
+	}
+
+	if count == 0 {
+		// flip the app config flag
+		appCfg, err := svc.ds.AppConfig(ctx)
+		if err != nil {
+			return ctxerr.Wrap(ctx, err, "retrieving app config")
+		}
+
+		appCfg.MDM.AppleBMEnabledAndConfigured = false
+		return svc.ds.SaveAppConfig(ctx, appCfg)
+	}
+
+	return nil
+}
+
+func (svc *Service) ListABMTokens(ctx context.Context) ([]*fleet.ABMToken, error) {
+	if err := svc.authz.Authorize(ctx, &fleet.AppleBM{}, fleet.ActionList); err != nil {
+		return nil, err
+	}
+
+	tokens, err := svc.ds.ListABMTokens(ctx)
+	if err != nil {
+		return nil, ctxerr.Wrap(ctx, err, "list ABM tokens")
+	}
+
+	return tokens, nil
+}
+
+func (svc *Service) UpdateABMTokenTeams(ctx context.Context, tokenID uint, macOSTeamID, iOSTeamID, iPadOSTeamID *uint) (*fleet.ABMToken, error) {
+	if err := svc.authz.Authorize(ctx, &fleet.AppleBM{}, fleet.ActionWrite); err != nil {
+		return nil, err
+	}
+
+	token, err := svc.ds.GetABMTokenByID(ctx, tokenID)
+	if err != nil {
+		return nil, ctxerr.Wrap(ctx, err, "get ABM token to update teams")
+	}
+
+	// validate the team IDs
+	var macOSTeamName, iOSTeamName, iPadOSTeamName string
+	if macOSTeamID != nil {
+		macOSTeam, err := svc.ds.Team(ctx, *macOSTeamID)
+		if err != nil {
+			return nil, ctxerr.Wrap(ctx, err, "checking existence of macOS team")
+		}
+
+		macOSTeamName = macOSTeam.Name
+	}
+
+	if iOSTeamID != nil {
+		iOSTeam, err := svc.ds.Team(ctx, *iOSTeamID)
+		if err != nil {
+			return nil, ctxerr.Wrap(ctx, err, "checking existence of iOS team")
+		}
+		iOSTeamName = iOSTeam.Name
+	}
+
+	if iPadOSTeamID != nil {
+		iPadOSTeam, err := svc.ds.Team(ctx, *iPadOSTeamID)
+		if err != nil {
+			return nil, ctxerr.Wrap(ctx, err, "checking existence of iPadOS team")
+		}
+		iPadOSTeamName = iPadOSTeam.Name
+	}
+
+	token.MacOSDefaultTeamID = macOSTeamID
+	token.IOSDefaultTeamID = iOSTeamID
+	token.IPadOSDefaultTeamID = iPadOSTeamID
+
+	if err := ctxerr.Wrap(ctx, svc.ds.SaveABMToken(ctx, token), "updating token teams"); err != nil {
+		return nil, ctxerr.Wrap(ctx, err, "updating token teams in db")
+	}
+
+	token.MacOSTeam = macOSTeamName
+	token.IOSTeam = iOSTeamName
+	token.IPadOSTeam = iPadOSTeamName
+
+	return token, nil
+}
+
+func (svc *Service) RenewABMToken(ctx context.Context, token io.Reader, tokenID uint) (*fleet.ABMToken, error) {
+	if err := svc.authz.Authorize(ctx, &fleet.AppleBM{}, fleet.ActionRead); err != nil {
+		return nil, err
+	}
+
+	oldTok, err := svc.ds.GetABMTokenByID(ctx, tokenID)
+	if err != nil {
+		return nil, ctxerr.Wrap(ctx, err, "get ABM token by id")
+	}
+
+	tok, err := svc.SaveABMToken(ctx, token)
```

Comment:
Hmm I'm not following, how does that update the same `tokenID`? That would create a new one, I think?

Reply:
@mna Ah I should have updated the name of `InsertABMToken`. It does an upsert actually, updating the `token` and `renew_at` columns.